engine/src/main/java/org/hibernate/search/query/engine/impl/HSQueryImpl.java

```diff
@@ -670,7 +680,51 @@ else if ( this.sort != null && projection != null ) {
 		);
 	}
 
+	private void validate(Sort sort, Collection<Class<?>> classes) {
+		SortField[] sortFields = sort.getSort();
+		for ( SortField sortField : sortFields ) {
+			String field = sortField.getField();
+			boolean indexedField = false;
+			for ( Class<?> clazz : classes ) {
+				IndexedTypeDescriptor indexedTypeDescriptor = extendedIntegrator.getIndexedTypeDescriptor( clazz );
+				FieldDescriptor fieldDescriptor = indexedTypeDescriptor.getIndexedField( field );
+				if ( fieldDescriptor != null ) {
+					indexedField = true;
+					if ( sortField instanceof DistanceSortField ) {
+						validateDistanceSortField( field, clazz );
+					}
+					break;
+				}
+			}
+			if ( !indexedField ) {
+				throw log.cannotSortBecauseFieldIsNotIndexed( field );
+			}
+		}
+	}
 
+	private void validateDistanceSortField(String field, Class<?> clazz) {
+		EntityIndexBinding indexBinding = extendedIntegrator.getIndexBinding( clazz );
+		PropertyMetadata propertyMetadata = indexBinding.getDocumentBuilder().getMetadata().getPropertyMetadataForProperty( field );
+		if ( propertyMetadata != null ) {
```

Comment:
Would it not make more sense to enhance the metadata itself so that the `PropertyMetadata` can already tell you whether you are dealing with coordinates? It seems wrong to work with reflection this late in the game. Also it is quite expensive.